### PR TITLE
Implement shared consumer #138

### PR DESCRIPTION
**Pull request by haneev**

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -1408,6 +1408,25 @@ func linkProperty(key string, value interface{}) LinkOption {
 	}
 }
 
+// LinkSourceName sets the name of the source queue
```
Comment (vcabbage) [inline client.go]: I'd like to see some detail added to this documentation, specifically:
* The link name must be unique on this connection.
* The default will be randomly generated.
* Sentences should end with punctuation. (https://github.com/golang/go/wiki/CodeReviewComments#comment-sentences)

For example:
```go
// LinkName sets the name of the link.
//
// The link names must be unique per-connection.
//
// Default: randomly generated.
``` 

Details about link names: http://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-transport-v1.0-os.html#doc-idp298752

File: client.go [hunk 2]
```diff
@@ -1408,6 +1408,25 @@ func linkProperty(key string, value interface{}) LinkOption {
...
+func LinkSourceName(name string) LinkOption {
+	return func(l *link) error {
+		l.name = name
```
Comment (vcabbage) [inline client.go]: I'd like to have some way of preventing duplicate names. It's not possible here, but it can be done during handle allocation.

You can check if the name already exists in the `linksByName` map and set an error on the link if it does (lines 523-525 are an example of setting an error). This should happen before `handles.next()` is called.

https://github.com/vcabbage/amqp/blob/61c744178624aae38c323c668e1a25a5995f6dd3/client.go#L519-L530

File: client.go [hunk 3]
```diff
@@ -1408,6 +1408,25 @@ func linkProperty(key string, value interface{}) LinkOption {
...
+		return nil
+	}
+}
+
+// LinkSourceCapabilities sets the source capabilities
```
Comment (vcabbage) [inline client.go]: Please add a period to the end of this sentence.

File: client.go [hunk 4]
```diff
@@ -1408,6 +1408,25 @@ func linkProperty(key string, value interface{}) LinkOption {
...
+func LinkSourceCapabilities(capabilities ...symbol) LinkOption {
```
Comment (vcabbage) [inline client.go]: I like to keep details like symbol vs string types internal to the lib. The capabilities here should be `...string` and converted to `symbol` inside the function. This will require a loop, but I think it makes for a nicer to use API.

File: client.go (hunk 2 above)
Comment (vcabbage) [reply client.go]: It would be nice to add a test for the duplicate name behavior, if you feel up to it. Don't feel obligated to do so though.

File: client_test.go [hunk 5]
```diff
@@ -80,4 +89,20 @@ func TestLinkOptions(t *testing.T) {
...
+	}
+
+	got, err := newLink(nil, nil, opts)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if got.name != expectedSourceName {
+		t.Errorf("Link Source Name does not match expected: %v got: %v", expectedSourceName, got.name)
+	}
```
Comment (vcabbage) [inline client_test.go]: Thanks for adding these tests! :100:

File: client.go (hunk 4 above)
Comment (haneev) [reply client.go]: Fixed

File: client.go (hunk 2 above)
Comment (haneev) [reply client.go]: Fixed

File: client.go (hunk 3 above)
Comment (haneev) [reply client.go]: Fixed

File: client.go [hunk 6]
```diff
@@ -1408,6 +1408,25 @@ func linkProperty(key string, value interface{}) LinkOption {
 	}
 }
 
+// LinkSourceName sets the name of the source queue
+func LinkSourceName(name string) LinkOption {
```
[the remark this answers is not in the document]
Comment (haneev) [reply client.go]: Fixed

File: README.md [hunk 7]
```diff
@@ -9,7 +9,7 @@
 
 pack.ag/amqp is an AMQP 1.0 client implementation for Go.
 
-AMQP 1.0 is not compatible with AMQP 0-9-1 or 0-10, which are
+[AMQP 1.0](http://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-complete-v1.0-os.pdf) is not compatible with AMQP 0-9-1 or 0-10, which are
```
[the remark this answers is not in the document]
Comment (haneev) [reply README.md]: Fixed

**Comment by vcabbage** [comment]
@haneev All the comments were marked fixed but I don't see the changes. Is there a commit that hasn't been pushed yet or is this just the general GitHub issues that's been happening for the past day or so?

**Comment by haneev** [comment]
> @haneev All the comments were marked fixed but I don't see the changes. Is there a commit that hasn't been pushed yet or is this just the general GitHub issues that's been happening for the past day or so?

Sorry, i marked them fixed in my branch. I have pushed the code now :)

File: client.go (hunk 2 above)
Comment (vcabbage) [reply client.go]: The current integration tests all rely on Microsoft Service Bus and Event Hubs. They're a bit of a pain to run unless you already have an Azure account that you don't mind giving the tests wide ranging access to. Some integration tests for other brokers that can be run in docker would be nice.

There are a couple issues with the test and I forgot that the entry in `linksByName` gets deleted after the attach returns. The fix isn't necessarily difficult, but since you're unable to run the tests it might be easier for me to resolve. Do you mind if I push some fixes to this PR? I might not get a chance to do it immediately since I'm at work and have a meeting to go to.

Comment (haneev) [reply client.go]: Thank you for helping me! I have updated this PR with your changes.

**Comment by haneev** [comment]
I have updated this PR with your changes. Thanks for helping!
